cpp/src/join/join_kernels.cuh

```diff
@@ -382,6 +424,124 @@ __global__ void probe_hash_table(multimap_type multi_map,
   }
 }
 
+/* --------------------------------------------------------------------------*/
+/**
+ * @brief  Performs a nested loop join to find all matching rows
+ between the left and right tables and generate the output for the desired Join operation.
+ *
+ * @param[in] left_table The left table
+ * @param[in] right_table The right table
+ * @param[in] check_row_equality The row equality comparator
+ * @param[in] left_num_rows The number of rows in the left table
+ * @param[in] right_num_rows The number of rows in the right table
+ * @param[out] join_output_l The left result of the join operation
+ * @param[out] join_output_r The right result of the join operation
+ * @param[in,out] current_idx A global counter used by threads to coordinate writes to the global
+ output
+ * @param[in] max_size The maximum size of the output
+ * @param[in] flip_results Whether to flip left and right outputs
+ * @param[in] offset An optional offset
+ * @tparam JoinKind The type of join to be performed
+ * @tparam block_size The number of threads per block for this kernel
+ * @tparam output_cache_size The side of the shared memory buffer to cache join output results
+ *
+ */
+/* ----------------------------------------------------------------------------*/
+template <join_kind JoinKind, cudf::size_type block_size, cudf::size_type output_cache_size>
+__global__ void nested_loop_join(table_device_view left_table,
+                                 table_device_view right_table,
+                                 row_equality check_row_equality,
+                                 const cudf::size_type left_num_rows,
+                                 const cudf::size_type right_num_rows,
+                                 cudf::size_type* join_output_l,
+                                 cudf::size_type* join_output_r,
+                                 cudf::size_type* current_idx,
+                                 const cudf::size_type max_size,
+                                 bool flip_results,
+                                 const cudf::size_type offset = 0)
+{
+  constexpr int num_warps = block_size / detail::warp_size;
+  __shared__ cudf::size_type current_idx_shared[num_warps];
+  __shared__ cudf::size_type join_shared_l[num_warps][output_cache_size];
+  __shared__ cudf::size_type join_shared_r[num_warps][output_cache_size];
+  cudf::size_type *output_l = join_output_l, *output_r = join_output_r;
```

Comment:
It looks like this was copied from `probe_hash_table`, _but_ I think this could be a bit cleaner. (if you prefer to not use `auto`, this would be `cudf::size_type * const&`)
```suggestion
  const auto& output_l = flip_results ? join_output_r : join_output_l;
  const auto& output_r = flip_results ? join_output_l : join_output_r;
```

Reply:
The `if (flip_result) {...}` below should be removed as well